src/RefractAPI.cc

```diff
@@ -195,116 +234,125 @@ namespace drafter {
+    refract::IElement* HeaderToRefract(const SectionInfo<snowcrash::Header>& header)
     {
         refract::MemberElement* element = new refract::MemberElement;
-        element->set(refract::IElement::Create(header.first), refract::IElement::Create(header.second));
+
+        element->set(refract::IElement::Create(header.section.first), refract::IElement::Create(header.section.second));
+
+        AttachSourceMap(element, header);
 
         return element;
     }
 
-    refract::IElement* AssetToRefract(const snowcrash::Asset& asset, const std::string& contentType, bool messageBody = true)
+    refract::IElement* AssetToRefract(const SectionInfo<snowcrash::Asset>& asset, const std::string& contentType, const std::string&  metaClass)
     {
-        if (asset.empty()) {
+        if (asset.section.empty()) {
             return NULL;
         }
 
-        refract::IElement* element = refract::IElement::Create(asset);
+        refract::IElement* element = PrimitiveToRefract(asset);
 
         element->element(SerializeKey::Asset);
-        element->meta[SerializeKey::Classes] = refract::ArrayElement::Create(messageBody ? SerializeKey::MessageBody : SerializeKey::MessageSchema);
+        element->meta[SerializeKey::Classes] = refract::ArrayElement::Create(metaClass);
 
         if (!contentType.empty()) {
+            // FIXME: has SourceMap?
             element->attributes[SerializeKey::ContentType] = refract::IElement::Create(contentType);
         }
 
         return element;
     }
 
-    refract::IElement* PayloadToRefract(const snowcrash::Payload* payload, const snowcrash::Action* action = NULL)
+    refract::IElement* PayloadToRefract(const SectionInfo<snowcrash::Payload>& payload, const SectionInfo<snowcrash::Action>& action)
     {
         refract::ArrayElement* element = new refract::ArrayElement;
         RefractElements content;
 
         // Use HTTP method to recognize if request or response
-        if (action == NULL || action->method.empty()) {
+        if (action.isNull() || action.section.method.empty()) {
             element->element(SerializeKey::HTTPResponse);
 
-            if (payload != NULL && !payload->name.empty()) {
-                element->attributes[SerializeKey::StatusCode] = refract::IElement::Create(payload->name);
+            if (!payload.isNull()) {
+                element->attributes[SerializeKey::StatusCode] = PrimitiveToRefract(MAKE_SECTION_INFO(payload, name));
             }
         }
         else {
             element->element(SerializeKey::HTTPRequest);
-            element->attributes[SerializeKey::Method] = refract::IElement::Create(action->method);
+            element->attributes[SerializeKey::Method] = PrimitiveToRefract(MAKE_SECTION_INFO(action, method));
 
-            if (payload != NULL) {
-                element->attributes[SerializeKey::Title] = refract::IElement::Create(payload->name);
+            if (!payload.isNull()) {
```

Comment:
Why?
- original code does not contain additional check. 
- current set of tests pass without part you require.

Are you able to deliver test to confirm your comment?